Add tab spacing modifier for tabs in TabBar and TabContainer

aaronjyoder opened the pull request:
Adds support for spacing between tabs.

- [x] LTR support
- [x] RTL support
- [x] Left/Center/Right Alignment support
- [x] Rearrange tab support
- [x] Clipping support
- [x] Allows both positive and negative values for design flexibility

Closes [proposal #10057](https://github.com/godotengine/godot-proposals/issues/10057).

KoBeWi commented:
The separation should probably be clamped, because it can be negative.
![image](https://github.com/user-attachments/assets/83bc0e91-8004-4ec0-a395-8a15e5326288)

aaronjyoder commented:
@KoBeWi Negative values are intended to be allowed, as at least one of the examples in the original proposal had a UI where negative values would be required for visual cohesion (think tabs with slanted/diagonal edges). What kind of clamp were you thinking?

I will look at the tab shifting issue--likely know what is causing that.

Should the tab separation also be considered for determining when to move tabs to the next page, or no? I didn't do it because it seemed out of scope and it causes lots of shifting when tabs move to a new page, but let me know what you think.

KoBeWi commented:
We usually do clamp separation to make sure it's non-negative:
https://github.com/godotengine/godot/blob/39c201ca58a8ddf1e99e62c8453294a2662a2fd8/scene/gui/button.cpp#L254
Though I guess the main reason is that it looks bad/glitched in most cases. Maybe allowing it is fine, if it works properly with clicks etc.

> Should the tab separation also be considered for determining when to move tabs to the next page, or no?

Separation pushes the tabs further, so this should be taken into account. Otherwise they could stick out of the panel etc. (I didn't test it thoroughly though)

aaronjyoder commented:
I'm struggling to figure out the correct way to account for pagination. If anyone has any insights, it would be appreciated.

```cpp
void TabBar::_update_cache(bool p_update_hover) {
	if (tabs.is_empty()) {
		buttons_visible = false;
		return;
	}

	int limit = get_size().width;
	int limit_minus_buttons = limit - theme_cache.increment_icon->get_width() - theme_cache.decrement_icon->get_width();

	int w = 0;

	max_drawn_tab = tabs.size() - 1;

	for (int i = 0; i < tabs.size(); i++) {
		tabs.write[i].text_buf->set_width(-1);
		tabs.write[i].size_text = Math::ceil(tabs[i].text_buf->get_size().x);
		tabs.write[i].size_cache = get_tab_width(i);

		tabs.write[i].truncated = max_width > 0 && tabs[i].size_cache > max_width;
		if (tabs[i].truncated) {
			int size_textless = tabs[i].size_cache - tabs[i].size_text;
			int mw = MAX(size_textless, max_width);

			tabs.write[i].size_text = MAX(mw - size_textless, 1);
			tabs.write[i].text_buf->set_width(tabs[i].size_text);
			tabs.write[i].size_cache = size_textless + tabs[i].size_text;
		}

		if (i < offset || i > max_drawn_tab) {
			tabs.write[i].ofs_cache = 0;
			continue;
		}

		tabs.write[i].ofs_cache = w;

		if (tabs[i].hidden) {
			continue;
		}

		w += tabs[i].size_cache;
		if (i < max_drawn_tab) { // Add separation between tabs
			w += theme_cache.tab_separation;
		}

		// Check if all tabs would fit inside the area.
		if (clip_tabs && i > offset && (w > limit || (offset > 0 && w > limit_minus_buttons))) {
			tabs.write[i].ofs_cache = 0;

			w -= tabs[i].size_cache;
			if (i < max_drawn_tab) {
				w -= theme_cache.tab_separation;
			}

			max_drawn_tab = i - 1;

			while (w > limit_minus_buttons && max_drawn_tab > offset) {
				tabs.write[max_drawn_tab].ofs_cache = 0;

				if (!tabs[max_drawn_tab].hidden) {
					w -= tabs[max_drawn_tab].size_cache;
					w -= theme_cache.tab_separation;
				}

				max_drawn_tab--;
			}
		}
	}

	missing_right = max_drawn_tab < tabs.size() - 1;
	buttons_visible = offset > 0 || missing_right;

	if (tab_alignment == ALIGNMENT_LEFT) {
		if (p_update_hover) {
			_update_hover();
		}
		return;
	}

	//int total_tab_separation = max_drawn_tab * theme_cache.tab_separation;
	if (tab_alignment == ALIGNMENT_CENTER) {
		//w = ((buttons_visible ? limit_minus_buttons : limit) - w - total_tab_separation) / 2;
		w = ((buttons_visible ? limit_minus_buttons : limit) - w) / 2;
	} else if (tab_alignment == ALIGNMENT_RIGHT) {
		//w = (buttons_visible ? limit_minus_buttons : limit) - w - total_tab_separation;
		w = (buttons_visible ? limit_minus_buttons : limit) - w;
	}

	for (int i = offset; i <= max_drawn_tab; i++) {
		tabs.write[i].ofs_cache = w;

		if (!tabs[i].hidden) {
			w += tabs[i].size_cache;
			if (i < max_drawn_tab) { // Add separation between tabs
                w += theme_cache.tab_separation;
            }
		}
	}

	if (p_update_hover) {
		_update_hover();
	}
}
```

That's what I'm trying right now. It works without pagination, but the problem is that as you increase `tab_separation`, it seems to remove tabs too early. I've spent several hours on it now and can't really figure out why, but maybe I am missing something simple.

KoBeWi commented:
You are adding and subtracting separation twice. There might be some case when the separation is counted twice without subtracting it.
The code would be more reliable if you added separation only once, without subtracting.

aaronjyoder commented:
Don't you need to subtract the separation, though, when you remove tabs? Otherwise it will still be considering separation as part of the total width without the tabs actually present. I just tried adding separation only once without subtracting and, as expected, it is incorrectly pruning tabs even earlier than before. Adding/subtracting only once has the same effect as the above code. Am I adding/subtracting in the wrong spot?

KoBeWi commented:
If you push the code with your attempted fix I'll see what's wrong.

KoBeWi commented:
btw another problem is that `tab_separation` is not accounted for minimum size when `clip_tabs` is disabled.

KoBeWi commented:
Ok I figured out why your code is wrong.
You are adding `tab_separation` *in front* of the tab. So e.g. in this situation: `[ 1 ] [ 2 ] [ 3 ]` the separation between tab 2 and 3 is treated as part of tab 2, so the tab 2 disappears when its front separation goes out of bounds. Also this condition
```
			w -= tabs[i].size_cache;
			if (i < max_drawn_tab) {
				w -= theme_cache.tab_separation;
			}
			max_drawn_tab = i - 1;
```
makes it worse, because when the tab 3 goes out of bounds, the buttons appear and tab 2 is clipped based on the extended size, so it disappears even earlier.

aaronjyoder commented:
> btw another problem is that `tab_separation` is not accounted for minimum size when `clip_tabs` is disabled.

Yeah I noticed that, I will fix it once I get the clipped stuff working. I think it may be a good idea to clamp the tab separation in 2 ways:

(1) Negative clamp to the minimum width of the TabBar container (so all buttons overlap each other, but no further)
(2) Positive clamp to the maximum width of the TabBar container.

This kind of clamp should fit all envisioned use-cases and prevent exiting the bounds of the TabBar container.

aaronjyoder commented:
With regard to your other comment, my understanding is this:
```
if (tabs[i].hidden) {
	continue;
}

w += tabs[i].size_cache;
if (i < max_drawn_tab) {
	w += theme_cache.tab_separation; // Prevents adding separation after last drawn tab--we still want to account for it otherwise
}

// Check if all tabs would fit inside the area.
if (clip_tabs && i > offset && (w > limit || (offset > 0 && w > limit_minus_buttons))) { // If you pass this check, you exceed the length, and have to clip a tab
	tabs.write[i].ofs_cache = 0;

	w -= tabs[i].size_cache;
	w -= theme_cache.tab_separation; // We're definitely drawnig at least one fewer tab, so remove the tab separation we just added

	max_drawn_tab = i - 1;

	while (w > limit_minus_buttons && max_drawn_tab > offset) { // remove tabs (and their width from accumulated width w) until we are below the width limit
		tabs.write[max_drawn_tab].ofs_cache = 0;

		if (!tabs[max_drawn_tab].hidden) {
			w -= tabs[max_drawn_tab].size_cache;
			w -= theme_cache.tab_separation; // We're repeatedly removing drawn tabs, so we have to remove their associated tab separation too
		}

		max_drawn_tab--;
	}
}
```
I put comments discussing my understanding. Is that understanding wrong?

aaronjyoder commented:
I figured it out.
```cpp
if (tabs[i].hidden) {
	continue;
}

w += tabs[i].size_cache;
if (i > 0) { // Don't add tab_separation for the first tab.
	w += theme_cache.tab_separation; // If i=1, we have 2 tabs, so add tab separation once. Going forward, this will add the correct amount of separation.
}

// Check if all tabs would fit inside the area.
if (clip_tabs && i > offset && (w > limit || (offset > 0 && w > limit_minus_buttons))) { // If you pass this check, you exceed the length, and have to clip a tab
	tabs.write[i].ofs_cache = 0;

	// We know i cannot be 0 if we are in here
	w -= tabs[i].size_cache;
	w -= theme_cache.tab_separation; // We're definitely drawnig at least one fewer tab, so remove the tab separation we just added

	max_drawn_tab = i - 1;

	while (w > limit_minus_buttons && max_drawn_tab > offset) { // remove tabs (and their width from accumulated width w) until we are below the width limit
		tabs.write[max_drawn_tab].ofs_cache = 0;

		if (!tabs[max_drawn_tab].hidden) {
			w -= tabs[max_drawn_tab].size_cache;
			w -= theme_cache.tab_separation; // We're repeatedly removing drawn tabs, so we have to remove their associated tab separation too
		}

		max_drawn_tab--;
	}
}
```
Pushing now.

aaronjyoder commented:
I will look at the other issues now:
(1) Enforcing a max width when `clip_tabs` is `false`.
(2) Buttons moving when they shouldn't as you scroll through pages.
(3) Button touch targets not being updated correctly (and are therefore incorrect, despite the buttons visually being correct)
(4) Setting a reasonable clamp that still allows for reasonable negative values.

If there is anything else, please let me know.

aaronjyoder commented:
Not quite sure what's wrong with CENTER/RIGHT alignments. It's drawing the unselected tabs perfectly, but the selected tab is not drawn correctly. It's drawn correctly for LEFT alignment, though.

KoBeWi commented:
For the shifting tabs problem, you can see here that the TabBar may display separation at the beginning:

https://github.com/user-attachments/assets/e21182e1-c6fe-45df-966c-e45fde27b527

This causes wrong tabs to be displayed and might be a clue on what is happening.

kitbdev commented:
I think some unit tests would be helpful.

>  Clamps range values when clipping is off such that tabs cannot exceed total container width

I don't think this is possible, since when clipping is off, the tab sizes determine the minimum width. Adding spacing will just increase the minimum width and there is no maximum size.

aaronjyoder commented:
> I don't think this is possible, since when clipping is off, the tab sizes determine the minimum width. Adding spacing will just increase the minimum width and there is no maximum size.

Ah, well then just incorporating the tab spacing into that calculation then, same difference, and ensuring that the negative tab spacing cannot shrink the tabs negatively beyond the minimum width.

KoBeWi commented:
It would be next to existing tests, e.g. `get_button.h`.
Though they could be probably added in a follow-up, seeing as this is your first pull request.

kitbdev commented:
There's a `tests/scene/test_tab_bar.h`, where it could be next to the `ensure tab visible` test case. 
Sometimes I find having tests helps me figure out the logic, but if you don't it's fine.

aaronjyoder commented:
I adjusted the code to expand the container if clipping is disabled. Seems to work well. However, I want to clamp `tab_separation` such that the most negative value is `get_tab_width()`. This would still allow overlapping tabs, but not allow them to extend beyond the edge of the container. Is there a standard way of doing this to ensure the display in-editor also abides by this? Or are you just meant to place a `MAX()` in there somewhere and allow the editor to display values that aren't being used?

aaronjyoder commented:
I've added drag-and-drop/rearrange support.

The previous implementation assumed you were always hovering over a tab, since tab separation was not supported. I have adjusted it so that if you are not hovering over a tab, it calculates the closest tab and continues from there as before.

The caret is now displayed at the midpoint between the two tabs you're currently between, instead of being on the left/right edge of a given tab.

aaronjyoder commented:
I am still having trouble figuring out clipping, but it is the last thing to do. Clipping currently works correctly and as-expected on the first page, but does not work correctly on subsequent pages.

aaronjyoder commented:
Alright, so I have tab clipping fully working with right alignment and center alignment. However, I cannot figure out how to get it to work with left alignment. It should just be a matter of figuring out the right math for `tab_separation_offset`. I messed around with it a bit but I need to give it a rest for now.

aaronjyoder commented:
Okay nvm I figured it out immediately after. Everything works now! Last thing to do would be a potential clamp, but let me know exactly how you might want to handle that. I could set some kind of clamp everywhere that I use `theme_cache.tab_separation` but that seems wrong to me.

AThousandShips commented:
Please avoid pushing so many times in a row, you currently require approval to run CI but if you had made contributions in the past you would have run CI several times

aaronjyoder commented:
Sorry about that, that makes sense.

aaronjyoder commented:
Alright, I think I've got it sorted now. I ran pre-commit locally before pushing, so hopefully that alleviates issues there.

I made an adjustment to the drag-and-drop rendering since it wasn't accounting for RTL correctly before, but should be now.

With regard to a clamp: I tested several other UI parameters, and it appears most of them do not have clamps of any kind, and let you go into the negatives with abandon, so I decided not to implement any kind of clamp in order to maintain consistency with that.

Calinou reviewed (approved):
Tested locally, it works as expected (including drag-and-drop rearranging). Code looks good to me.

**Testing project:** [test_pr_103214.zip](https://github.com/user-attachments/files/19063001/test_pr_103214.zip)

![image](https://github.com/user-attachments/assets/455aae9f-4149-49ec-b117-987e95eeea95)

Note that clicking empty spaces between tabs won't select any tabs, but it'll focus the TabContainer/TabBar. This also occurred when you click space at the right of the last tab before this PR.

I agree with not preventing negative spacing, as we allow it for other theme constants like line spacing.

AThousandShips commented on scene/gui/tab_bar.cpp:
```diff
@@ -1398,6 +1446,24 @@ int TabBar::get_tab_idx_at_point(const Point2 &p_point) const {
+int TabBar::get_closest_tab_idx_to_point(const Point2 &p_point) const {
+	int closest_tab = get_tab_idx_at_point(p_point); // See if we're hovering over a tab first
+
+	if (closest_tab == -1) { // Didn't find a tab, so get the closest one
```
```suggestion
	if (closest_tab == -1) { // Didn't find a tab, so get the closest one.
```

KoBeWi commented:
Separation still has problems:

https://github.com/user-attachments/assets/13d3f936-71e5-4dad-b062-e469e19234b6

There is no wrong offsets, but the TabBar should display 2 tabs on every page.

aaronjyoder commented:
There is a regression that was introduced at some point where tabs on other pages are not obeying center/right alignment. I swear I checked that just about every single commit but nonetheless. Looking into it now.

aaronjyoder commented:
Regression fixed. I am unsure how best to ensure 2 tabs are always visible, but I'll think on it for a bit.

aaronjyoder commented:
Whoops! Looks like fixing that other regression also allows me to remove the special exception for how left alignment is handled around line 400. :) Fixed locally, will wait to push until I implement a solution for the 2 tab on each page requirement.

KoBeWi commented:
> until I implement a solution for the 2 tab on each page requirement.

I already mentioned that it seems to be fixed already (last time I tested). Can you still reproduce it?

aaronjyoder commented:
> Can you still reproduce it?
Well, yes, if you increase separation to a very large value, you can ensure there is only 1 tab on each page, rather than 2. Is that not what you meant?

aaronjyoder committed:
Add tab spacing modifier for tabs in TabBar and TabContainer

KoBeWi commented:
I meant that there was less tabs per page than could fit. If separation is big enough that no other tab can be added then it's fine.
Sorry if I was unclear, I didn't mean that there should be strictly 2 tabs. I was only referring to the case shown in my video.

KoBeWi reviewed (approved):
Looks good now.

Repiteo commented:
Thanks! Congratulations on your first merged contribution! 🎉

aaronjyoder commented:
> @aaronjyoder Any interest in implementing this proposal? [godotengine/godot-proposals#8397](https://github.com/godotengine/godot-proposals/discussions/8397) You were already most of the way there, you just need an "Expand Tabs to Fill" toggle.

@taelspinner Yeah I can take a look at it this week.

aaronjyoder commented:
@COTE-Allan I've got a PR for it [here](https://github.com/godotengine/godot/pull/113385). We should direct all further discussion here, as I'm sure the maintainers don't want unrelated discussion on a now-merged PR. :)
